internal/rootcoord/root_coord.go

```diff
@@ -2992,3 +3049,445 @@ func (c *Core) ListCredUsers(ctx context.Context, in *milvuspb.ListCredUsersRequ
+				errMsg := "fail to remove user from role"
+				logger.Error(errMsg, zap.String("role_name", roleResult.Role.Name), zap.String("username", userEntity.Name), zap.Int("current_index", index), zap.Error(err))
+				return failStatus(commonpb.ErrorCode_OperateUserRoleFailure, errMsg), err
+			}
+		}
+	}
+	if err = c.MetaTable.DropRole(util.DefaultTenant, in.RoleName); err != nil {
+		errMsg := "fail to drop the role"
+		logger.Error(errMsg, zap.String("role_name", in.RoleName), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_DropRoleFailure, errMsg), err
+	}
+
+	logger.Debug(method+" success", zap.String("role_name", in.RoleName))
+	metrics.RootCoordDDLReqCounter.WithLabelValues(method, metrics.SuccessLabel).Inc()
+	metrics.RootCoordDDLReqLatency.WithLabelValues(method).Observe(float64(tr.ElapseSpan().Milliseconds()))
+	metrics.RootCoordNumOfRoles.Dec()
+	return succStatus(), nil
+}
+
+// OperateUserRole operate the relationship between a user and a role
+// - check the node health
+// - check if the role is valid
+// - check if the user is valid
+// - operate the user-role by the metatable api
+// - update the policy cache
+func (c *Core) OperateUserRole(ctx context.Context, in *milvuspb.OperateUserRoleRequest) (*commonpb.Status, error) {
+	method := "OperateUserRole-" + in.Type.String()
+	metrics.RootCoordDDLReqCounter.WithLabelValues(method, metrics.TotalLabel).Inc()
+	tr := timerecord.NewTimeRecorder(method)
+	logger.Debug(method, zap.Any("in", in))
+
+	if code, ok := c.checkHealthy(); !ok {
+		return errorutil.UnhealthyStatus(code), errorutil.UnhealthyError()
+	}
+
+	if _, err := c.MetaTable.SelectRole(util.DefaultTenant, &milvuspb.RoleEntity{Name: in.RoleName}, false); err != nil {
+		errMsg := "fail to check the role name"
+		logger.Error(errMsg, zap.String("role_name", in.RoleName), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_OperateUserRoleFailure, errMsg), err
+	}
+	if _, err := c.MetaTable.SelectUser(util.DefaultTenant, &milvuspb.UserEntity{Name: in.Username}, false); err != nil {
+		errMsg := "fail to check the username"
+		logger.Error(errMsg, zap.String("username", in.Username), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_OperateUserRoleFailure, errMsg), err
+	}
+	if err := c.MetaTable.OperateUserRole(util.DefaultTenant, &milvuspb.UserEntity{Name: in.Username}, &milvuspb.RoleEntity{Name: in.RoleName}, in.Type); err != nil {
+		errMsg := "fail to operate user to role"
+		logger.Error(errMsg, zap.String("role_name", in.RoleName), zap.String("username", in.Username), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_OperateUserRoleFailure, errMsg), err
+	}
+
+	var opType int32
+	if in.Type == milvuspb.OperateUserRoleType_AddUserToRole {
+		opType = int32(typeutil.CacheAddUserToRole)
+	} else if in.Type == milvuspb.OperateUserRoleType_RemoveUserFromRole {
+		opType = int32(typeutil.CacheRemoveUserFromRole)
+	}
```

Comment:
because there are only two types, maybe it isn't mush difference.

Reply:
Yes. Logical equivalent. `switch` actually mention us to handle the case of `default`, which I missed in last review.

Reply:
Because `in.Type` has been checked in the proxy server, I think in the root coord serve the params must be valid. If checking them twice, there will be many unnecessary codes.